lib/open_project/authentication/strategies/warden/basic_auth_failure.rb

```diff
@@ -0,0 +1,16 @@
+module OpenProject
+  module Authentication
+    module Strategies
+      module Warden
+        ##
+        # This strategy is inserted after optional basic auth strategies to
+        # indicate that invalid basic auth credentials were provided.
+        class BasicAuthFailure < ::Warden::Strategies::BasicAuth
+          def authenticate_user(_username, _password)
+            nil # always fails
```

Comment:
Okay, I think now I've got it. In my own words:


Whenever a stragegy is valid it decides *ultimately* over grant or denial, that means the next strategy in the list is **only** tried when the current one was **not** valid.

This means without Basic auth headers, all (three) basic auth strategies are invalid, thus `session` has to decide.

With basic auth headers it depends:
At least the fail strategy is hit, so we don't fall back to session, because the fail strategy comes first (but only when HTTP basic auth headers are present).
In the special cases `user == 'apikey'` or `user == 'what_was_configured'` the other strategies are valid, but will decide *ultimately* themselves.

That means **this** strategy only denies access when:

* HTTP basic auth is used
* none of the other basic auth strategies hit

That also means, if this strategy would be missing:

* fallback would be the session strategy (or no strategy under different configuration circumstances)
* only if session would recognize you, you would **not** get a 401
    * i.e. no backdoor would ever be introduced alone by forgetting this strategy (this is what I feared in the first place)